src/vmm/src/persist.rs

```diff
@@ -295,6 +312,19 @@ fn guest_memory_from_file(
     GuestMemoryMmap::restore(&mem_file, mem_state, track_dirty_pages).map_err(DeserializeMemory)
 }
 
+fn validate_devices_number(
+    device_states: &DeviceStates,
+) -> std::result::Result<(), CreateSnapshotError> {
+    use self::CreateSnapshotError::TooManyDevices;
+    let devices_number = device_states.block_devices.len()
+        + device_states.net_devices.len()
+        + device_states.vsock_device.as_ref().map_or(0, |_| 1);
```

Comment:
This is missing `Balloon` (and on `aarch64` it is missing RTC and Serial).

To make this code less error prone (like forgetting to add some device type or smth), I suggest creating a `device_manager::devices_count()` and contain device management (like keeping count) to `device_manager`

Reply:
You are right, this needs a change.
I surfed the code a bit and it seems to me that the `MMIODeviceManager` already keeps track of all devices attached using a `id_to_dev_info` hashmap. I thought it would be a good idea to use that one to fetch the devices number. WDYT?